forgotten-sender/listing-04-fix-2/main.go

```diff
@@ -0,0 +1,70 @@
+// Example program with the goroutine leak fixed. The code cancels the worker
+// on its way out of the process function.
+package main
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"time"
+)
+
+func main() {
+
+	// Report number of goroutines. Will be 1.
+	fmt.Println("Number of goroutines:", runtime.NumGoroutine())
+
+	process(context.Background())
+
+	// Hold the program from terminating for 1 second to see if the goroutine
+	// created by process will terminate.
+	time.Sleep(time.Second)
+
+	// Report number of goroutines. Will be 1.
+	fmt.Println("Number of goroutines:", runtime.NumGoroutine())
+}
+
+func process(ctx context.Context) {
+
+	// Wrap the parent context in a new context with a 100ms deadline. The
+	// context will be canceled if 100ms passes or if the parent function's
+	// context is canceled for any reason.
+	ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+	defer cancel()
+
+	// Make a channel for our goroutine to report its result.
+	ch := make(chan string)
+
+	// Start a worker to do some work then either send on the channel or abort if
+	// the context was canceled.
+	go func() {
+
+		// Get the result of the work. Ideally ctx could be passed down into
+		// doSomeWork which would know how to cancel early. Realistically, in many
+		// cases it is not cancellable and the worker must wait for it to finish.
+		result := doSomeWork()
```

Comment:
Back when we were passing ctx down I had the call to `doSomeWork` right in line with the select case.

Now I think we should take it out. If we can't cancel `doSomeWork` then no matter what this goroutine won't return until `doSomeWork` returns. If we embed it directly in the select block I worry people will think it can be canceled early; but it can't.

Reply:
This goroutine should always look the same.

```
// Start a worker to do some work then send on the channel.
go func() {
	ch <- doSomeWork()
	fmt.Println("Worker terminated")
}()
```